[clang-tidy] fix misc-const-correctnes false-positive for fold expressions

5chmidti opened the pull request:
The check no longer emits a diagnostic for non-parameter-pack 
variables in C++17 fold expressions.
The operator used is type-dependent because of the parameter pack
and can therefore not be guaranteed to not mutate the variable.

Fixes: #70323

5chmidti committed:
[clang-tidy] fix misc-const-correctnes false-positive for fold expressions

The check no longer emits a diagnostic for variables used as the
initializer of C++17 fold expressions.
The operator used is type-dependent because of the parameter pack
and can therefore not be guaranteed to not mutate the variable.

Fixes: #70323

PiotrZSL reviewed (approved):
LGTM, but i'm not 100% sure about ExprMutationAnalyzer part (if thats a right place).
Leave it open for few days, so others could comment if they have objections.

5chmidti committed:
use -fno-delayed-template-parsing in added test in ExprMutationAnalyzerTest

HerrCai0907 commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -382,7 +382,8 @@ Changes in existing checks
   type-dependent binary operator. Improved performance of the check through
-  optimizations.
+  optimizations. The check no longer emits a diagnostic for
+  variables used as the initializer of C++17 fold expressions.
```
I am not sure "initializer of C++17 fold expressions" is the correct description for `InitExpr` in fold expression since I don't find it in https://eel.is/c++draft/expr.prim.fold and https://en.cppreference.com/w/cpp/language/fold.

5chmidti replied:
You are right, what do you think about `The check no longer emits a diagnostic for non-parameter-pack variables in C++17 fold expressions.`?

5chmidti replied:
Done

5chmidti committed:
improve and fix tests

5chmidti commented:
I fixed the clang-tidy test case by separating them and I have added a test case to check that using a variable in the pack side of the fold expression would be detected (that did not change with this patch though, just as a sanity check).

5chmidti committed:
better change description in release note

PiotrZSL commented:
@PiotrZSL I would say: fell free to merge it. It's not that it would break too much. In worst case we got around ~1 month to fix it on a branch (if someone would complain).

PiotrZSL commented:
@5chmidti I would say: fell free to merge it. It's not that it would break too much. In worst case we got around ~1 month to fix it on a branch (if someone would complain).
